src/Checkpoint/CheckpointRequest.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <Checkpoint/CheckpointContext.h>
+#include <Checkpoint/CheckpointContextFwd.h>
+
+#include <base/defines.h>
+
+#include <atomic>
+#include <cassert>
+#include <mutex>
+
+namespace DB
+{
+/// \CheckpointRequest applies double check locking to minimize using mutex
+/// directly for better performance
+class CheckpointRequest
+{
+public:
+    /// \poll moves out the existing checkpoint context request atomically if there is any.
+    CheckpointContextPtr poll() noexcept
+    {
+        CheckpointContextPtr new_request;
+        if (has_ckpt_request)
+        {
+            std::scoped_lock lock(ckpt_mutex);
+            new_request.swap(ckpt_ctx);
+            has_ckpt_request.store(false);
+        }
+        return new_request;
+    }
+
+    void setCheckpointRequestCtx(CheckpointContextPtr ckpt_ctx_) noexcept
+    {
+        assert(!has_ckpt_request);
+        /// Perform a deep copy to ensure that the checkpoint file is stored in a distinct path.
+        CheckpointContextPtr new_ckpt = std::make_shared<CheckpointContext>(*ckpt_ctx_);
```

Comment:
I don't think we will need a deep copy ?

Reply:
it works with shallow copy today.. but I remembered yesterday if I do not take the deep copy, the smoke test will break.
and I find the original atomic<ckpt_ctx*> also does deep copy, thus I take this trial.